### ISSUE TITLE
Problem with plural rules

### ISSUE THREAD

**Issue by nbouvrette** [formatjs/formatjs-old#185]
**Which package?**
https://github.com/formatjs/formatjs/tree/master/packages/intl-messageformat

**Describe the bug**
Before I open up a bug with Chromium, I would like to confirm my finding is correct.

I noticed a few issues when trying to parse plural ICU messages and noticed that `intl-messageformat` uses the Browser plural rules which I suspect in Chrome are not aligned with the CLDR.

**To Reproduce**

In Chrome console:

```js
console.log((new IntlMessageFormat.IntlMessageFormat('Test plural form: {count, plural, zero {zero} one {one} two {two} few {few} many {many} other {other}} | value: {count}', 'as')).format({count: '0'}));
```

Output:

`Test plural form: other | value: 0`

Now if I do this in Chromium:

```js
new Intl.PluralRules('as').select(0);
```

Output:

```js
"other"
```

Now if I do this in Firefox (output):

```js
"one"
```

**Expected behavior**
The expected behavior for this case is the one Firefox is showing:

 ```js
"one"
```

**Screenshots**
Not applicable.

**Desktop (please complete the following information):**

- OS: OSX latest
- Browser Chrome latest, Firefox latest
- Version [e.g. 22] latest

**Additional context**
Please confirm that this is a Chromium issue and I will open the issue with them.

**Comment by longlho** [comment]
Yes CLDR data gets released twice a year and browsers ship with specific
versions so they will get out of date.

**Comment by nbouvrette** [comment]
Sounds good - I saw other issues with a few other few rules. I'm not sure if you are familiar with [CLDR](http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html#en) rules but if you are can you confirm this:

Rule for `one` English: `i = 1 and v = 0`

If I interpret this correctly based on the [plural rules](https://www.unicode.org/reports/tr35/tr35-numbers.html#Language_Plural_Rules), this would allow only the number `1`, excluding any trailing zeros, which means that `1.0` or should apply the `other` rule?

This is both incorrect in Chromium and Firefox:

`new Intl.PluralRules('en').select('1.00');`

Output:

`"one"`

Happy to also add this in the issue if you think this is also not expected but I'm a bit surprised since it impacts English.

**Comment by longlho** [comment]
Yeah that looks incorrect based on strict operand definition. Last thing is to check if ICU4J does it right

**Comment by longlho** [comment]
So based on the spec, `1.00` yields `one` is correct: https://tc39.es/ecma402/#sec-intl.pluralrules.prototype.select

So it'd basically be `'1.00'` -> `1 as number`, then trigger `GetOperands` (https://tc39.es/ecma402/#sec-resolveplural) and use that to do `PluralRuleSelect`

For ICU4J, for regular `.select(double)`, there's no way to pass in `1.00`. The only way to do that is to pass in the result from a `NumberFormatter`, and that does give the correct output of `other`
```java
PluralRules.forLocale(locale).select(
	NumberFormatter
		.withLocale(locale)
		.precision(Precision.fixedFraction(3))
		.format(1)
) 
```

**Comment by longlho** [comment]
ok so there is a way to do that, by specifying `minimumFractionDigits`: https://github.com/tc39/ecma402/issues/365

MDN is outdated though

**Comment by longlho** [comment]
I'll keep this open since this might be a bug in our polyfill as well

**Comment by nbouvrette** [comment]
Sorry, I wish I had more time to help you investigate this one yesterday but it looks like you got everything figured out :) thanks for resolving this and finding the missing info!

By the way, how is the Polyfill triggered right now? should it check that not only `Intl` is available but also that `minimumFractionDigits` works? From a quick check, it looks like you could overwrite the native browser object object with the polyfill if the native browser support is not adequate to respect the plural rules.

**Comment by longlho** [comment]
So as of the issue above, FF & Chrome 77 support it. In general it's a fairly edge case and technically I can do bug detection to overwrite, but I'm not sure if it's worth the library weight tradeoffs.

**Comment by nbouvrette** [comment]
I agree about the edge case part but keeping my current project in mind, a visual tool which should be able to explain rules in "human understandable language" (not ICU rules) and having the ability to test out variables to check the result, I am a bit worried that some user will have older browsers and get inconsistent results which might impact their trust in the tool.

Having said this, maybe I could implement a more rigid Polyfill loader on my end to make sure the experience is consistent across browsers.

**Comment by longlho** [comment]
yeah we do dist out raw polyfill file w/o the env check so u can write your own env checker

**Comment by nbouvrette** [comment]
I just tried to write my own environment checker but I could not find a UMD version for the polyfill. 

Is this the right NPM: https://www.npmjs.com/package/intl-locales-supported

Did I miss something?

**Comment by longlho** [comment]
https://www.npmjs.com/package/@formatjs/intl-pluralrules
https://www.npmjs.com/package/@formatjs/intl-relativetimeformat

**Comment by nbouvrette** [comment]
Thanks - the polyfill works well. It actually works just like the native one. I was hoping it would not reproduce this new issue I opened here:

https://github.com/tc39/ecma402/issues/373

I was planning to propose to try to do a PR if you think it should be fixed, but I'm not sure I fully understand why `toRawFixed` was implemented that way and also how to deal with `1.9999999999999999`. 

Even this number seems to break ICU4J, but maybe this is starting to be a bit too much in the realm of an edge case that will never be triggered. 

I'm wondering if truncating might not be a better approach to match the CLDR, but I know there are also some rules that factor in trailing zeros (e.g. Icelandic).

**Comment by longlho** [comment]
We use test262 for our polyfills so it behaves exactly like browsers.
ToRawFixed was a stright implementation from the spec if you look it up. As
Shane mentioned in that issue default truncation is 3 so that’s why you see
that issue. Changing maximumFractionDigits would effectively fix the
problem.

It is very edge yeah. There aren’t a whole lot of languages dealing with
fractions, let alone more than 3.

**Comment by nbouvrette** [comment]
I just checked and French, Portuguese, and Slovak are amongst languages that could be impacted by this behavior. That is a lot of people.

I think if the behavior would match ICU4J, at least the edge case would have less chance to be triggered than after 3 digits. 

We'll see if the change gets approved in the other issue!

**Comment by nbouvrette** [comment]
I closed my other issue regarding the 3 digit rounding - looks like it would cause too much disruption. 

I saw that you changed one of the polyfill files I was using (probably should have been using another file):

`Renamed @formatjs/intl-pluralrules/dist/polyfill-with-locales-for-test262.js` (https://github.com/formatjs/formatjs/commit/6bc0bd7)

I was thinking that the UMD folder would contain a polyfill for browsers that would also include all languages and could not find one there. This is why I started using the other file which I suspect is less backward compatible.

Am I using the right file or should there be a `with-locales` version in the UMD folder?
